### ISSUE TITLE
question of weighted_mask_loss function

### ISSUE THREAD

**Issue by shuyanzhou** [askforalfred/alfred#16]
Hi,

The weighted_mask_loss function inside seq2seq_im_mask looks like:
```python
    def weighted_mask_loss(self, pred_masks, gt_masks):
        '''
        mask loss that accounts for weight-imbalance between 0 and 1 pixels
        '''
        # pred_mask [batch * len, 1, 300, 300], gt_mask[batch * len, 1, 300, 300]
        bce = self.bce_with_logits(pred_masks, gt_masks)
        flipped_mask = self.flip_tensor(gt_masks)
        inside = (bce * gt_masks).sum() / (gt_masks).sum()
        outside = (bce * flipped_mask).sum() / (flipped_mask).sum()
        return inside + outside
```
However, the summation is *batch-wise* instead of *sample-wise*, which might be problematic. Besides, I was wondering whether it is necessary to have an average. Here is the revised code based on my understanding:
```python
    def weighted_mask_loss(self, pred_masks, gt_masks):
        '''
        mask loss that accounts for weight-imbalance between 0 and 1 pixels
        '''
        # pred_mask [batch * len, 1, 300, 300], gt_mask[batch * len, 1, 300, 300]
        bce = self.bce_with_logits(pred_masks, gt_masks)
        flipped_mask = self.flip_tensor(gt_masks)
        bs = bce.shape[0]
        bce = bce.view(bs, -1)
        gt_masks = gt_masks.view(bs, -1)
        flipped_mask = flipped_mask.view(bs, -1)
        # instant-wised summation
        inside = torch.sum(bce * gt_masks, dim=1) / torch.sum(gt_masks,dim=1)
        outside = torch.sum(bce * flipped_mask, dim=1) / torch.sum(flipped_mask, dim=1)
        per_loss = inside + outside
        # average
        loss = torch.mean(per_loss)
        return loss, per_loss
```

Is my understanding correct? Thanks!

**Comment by MohitShridhar** [comment]
Thanks! Your loss function makes sense. Does it produce lower mask loss on val scenes?

**Comment by shuyanzhou** [comment]
@MohitShridhar Thanks for the reply. I haven't test yet, will update once I have results.

**Comment by MohitShridhar** [comment]
Super old issue. Closing for now.
